contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+3. Clear the `VolumeAttachment.Status.AttachError` field.
+
+If the operation fails, the external-attacher will:
+
+1. Ensure the `VolumeAttachment.Status.IsAttached` field to still false to indicate the volume is not attached.
+2. Set the `VolumeAttachment.Status.AttachError` field detailing the error.
+3. Create an event against the Kubernetes API associated with the `VolumeAttachment` object to inform users what went wrong.
+
+The external-attacher may implement it’s own error recovery strategy, and retry as long as conditions specified for attachment above are valid. It is strongly recommended that the external-attacher implement an exponential backoff strategy for retries.
+
+The detach operation will be triggered by the deletion of the `VolumeAttachment` Kubernetes API objects. Since the `VolumeAttachment` Kubernetes API object will have a finalizer added by the external-attacher, it will wait for confirmation from the external-attacher before deleting the object.
+
+Once all the following conditions are true, the external-attacher should call `ControllerUnpublishVolume` against the CSI volume driver to detach the volume from the specified node:
+1. A `VolumeAttachment` Kubernetes API object is deleted by Kubernetes: the value for the `VolumeAttachment.metadata.deletionTimestamp` field is set.
+
+If the operation completes successfully, the external-attacher will:
+1. Remove its finalizer from the list of finalizers on the `VolumeAttachment` object permitting the delete operation to continue.
+
+If the operation fails, the external-attacher will:
+
+1. Ensure the `VolumeAttachment.Status.IsAttached` field remains true to indicate the volume is not yet detached.
+2. Set the `VolumeAttachment.Status.DetachError` field detailing the error.
+3. Create an event against the Kubernetes API associated with the `VolumeAttachment` object to inform users what went wrong.
+
+The new API object called `VolumeAttachment` will be defined as follows:
+
+```GO
+// VolumeAttachment captures the intent to attach or detach the specified
+// volume to/from the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+  metav1.TypeMeta `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+  // Specification of the desired attach/detach volume behavior.
+  Spec VolumeAttachmentSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Most recently observed status of the VolumeAttachment request.
+  // This data may not be up to date.
+  Status VolumeAttachmentStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+
+type VolumeAttachmentSpec struct {
+  // Attacher indicates the name of the volume plugin that MUST handle this
+  // request. This is the name returned by GetPluginName().
+  Attacher string `json:"attacher" protobuf:"bytes,1,opt,name=attacher"`
```

Comment:
Could clarify if this is the driver we are talking about or the proxy container.

Reply:
Clarified above.